docs/src/lens.md

```diff
@@ -1,149 +1,251 @@
 # Lenses and the `Lens` trait
 
-Let's say we're building a todo list application, and we are designing the widget
-that will represent a single todo item. Our data model looks like this:
+One of the key abstractions in `druid` along with `Data` is the `Lens` trait. This page explains what they are, and then how to use them. `Lens`es may seem complicated at first, but they are also very powerful, allowing you to write code that is reusable, concise, and understandable (once you understand `Lens`es themselves).
 
-```rust,noplaypen
-{{#include ../book_examples/src/lens_md.rs:todo_item}}
+## Fundamentals: Definition and Implementation
+
+Like Rust itself, lenses are one of those things that require effort up front to learn, but are very fun and effective to use once you understand them. This section represents the effort part of the equation. I promise if you stick with it you will reap the rewards.
+
+### Definition
+
+Let's start with the definition of a `Lens`:
+
+```rust
+pub trait Lens<T, U> {
+    fn with<F: FnOnce(&U)>(&self, data: &T, f: F);
+
+    fn with_mut<F: FnOnce(&mut U)>(&self, data: &mut T, f: F);
+}
 ```
 
-We would like our widget to display the title of the item, and then below
-that to display two checkmarks that toggle the 'completed' and 'urgent' bools.
-`Checkbox` (a widget included in Druid) implements `Widget<bool>`.
-How do we use it with `TodoItem`? By using a `Lens`.
+I've copied this definition from the `druid` source code, but then simplified it a little, by removing the return types, as they are not fundamental to the way lenses work.
 
-## Conceptual
+The first thing to notice is the generics on the `Lens` itself. There are 3 types involve in the lens: the lens itself, `T` and `U`. The two type parameters represent the mis-match that lenses solve: we have a function that operates on `U`, and an object of type `T`, so we need to transform `T` into `U` somehow.
 
-You can think of a lens as a way of "focusing in" on one part of the data. You
-have a `TodoItem`, but you *want* a `bool`.
+### Implementation
 
-`Lens` is a trait for types that perform this "focusing in" (aka *lensing*).
-A simplified version of the `Lens` trait might look like this:
+Time for an example. Let's implement & use `Lens` manually so we can see what's going on.
 
-```rust,noplaypen
-{{#include ../book_examples/src/lens_md.rs:simple_lens}}
+```rust
+struct Container {
+    inner: String,
+    another: String,
+}
+
+// Here the lens doesn't have any data, but there are cases where
+// it might, for example it might contain an index into a collection.
+struct InnerLens;
+
+// Our lens will apply functions that operate on a `String` to a `Container`.
+impl Lens<Container, String> for InnerLens {
+    fn with<F: FnOnce(&String)>(&self, data: &Container, f: F) {
+        f(&data.inner);
+    }
+
+    fn with_mut<F: FnOnce(&mut String)>(&self, data: &mut Container, f: F) {
+        f(&mut data.inner);
+    }
+}
```

Comment:
So, now `InnerLens` can be renamed to `NameLens`. Part of the issue I had reading this initially, was that I thought InnerLens meant lens that was internal rather than a lens that lenses into `inner`. I just kept thinking inner was something else. This is my fault, but think it would be better to have more memorable field names for the example.

Reply:
Also maybe there should be an explanation of the paramter `f: F`. What is that function doing under the hood?